git: implement a next-generation git client

stevekuznetsov opened the pull request:
The current client in `prow/git` is often esoteric in how it is set up
and is not unit-testable. The `prow/git/v2` packge aims to re-implement
the original client while focusing on:

 - dynamically loading and resolving when input is not static, like:
   - authentication tokens
   - usernames, when derived from an API
 - testability at every level, especially the unit level
 - factoring between repository clients that expect a cloned directory
   and operate on it versus clients that use a central cache

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

/assign @cjwagner @alvaroaleman @fejta

stevekuznetsov commented:
TODO:

 - [ ] refactor the `git` integration tests to expect an interface for the client the use, not a literal
 - [ ] show that the `v2` client works in those integration tests
 - [ ] use the new feature end-to-end in a component that does not use the `v1` client, like the autobumper

fejta commented:
Do you want me to review this now or after the TODOs are done or after its not WIP?

alvaroaleman commented on prow/flagutil/git.go:
```diff
@@ -0,0 +1,90 @@
+	fs.StringVar(&o.user, "git-user", "", "User for git commits, optional. Can be derived from GitHub credentials.")
+	fs.StringVar(&o.email, "git-email", "", "Email for git commits, optional. Can be derived from GitHub credentials.")
+	fs.StringVar(&o.tokenPath, "git-token-path", "", "Path to the file containing the git token for HTTPS operations, optional. Can be derived from GitHub credentials.")
+	fs.BoolVar(&o.useSSH, "git-over-ssh", false, "Use SSH when pushing and pulling instead of HTTPS. SSH credentials should be present at ~/.ssh")
```
What do we want to do re SSH  host keys, leave the responsibility on the one building the image?

alvaroaleman commented on prow/git/v2/executor_test.go:
```diff
@@ -0,0 +1,113 @@
+		{
+			name: "happy path with nothing to censor returns all output",
+			dir:  "/somewhere/repo",
+			git:  "/usr/bin/git",
```
Nit: When you add something like the `git` binary to each test case I expect a testcase that does have it set differently. Maybe just remove here and hardcode to `/usr/bin/git` for these tests?

alvaroaleman commented on prow/git/v2/interactor.go:
```diff
@@ -0,0 +1,256 @@
+	RevParse(commitlike string) (string, error)
+	BranchExists(branch string) bool
+	CheckoutNewBranch(branch string) error
+	Merge(commitlike string) (bool, error)
```
Would it make sense to move the merge-related bits into the publisher so they can profit from the `info` for their author configuration?

alvaroaleman commented on prow/git/v2/remote.go:
```diff
@@ -0,0 +1,101 @@
+// for the repository.
+func (f *simpleAuthResolverFactory) CentralRemote(org, repo string) RemoteResolver {
+	return simpleAuthResolver(func() (*url.URL, error) {
+		return &url.URL{Scheme: "https", Host: f.host, Path: fmt.Sprintf("%s/%s", org, repo)}, nil
```
I believe the path needs a `.git` suffix

stevekuznetsov replied:
Happens to work either way:

```
$ git clone https://github.com/kubernetes/test-infra
Cloning into 'test-infra'...
remote: Enumerating objects: 2, done.
remote: Counting objects: 100% (2/2), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 203116 (delta 1), reused 0 (delta 0), pack-reused 203114
Receiving objects: 100% (203116/203116), 96.02 MiB | 4.40 MiB/s, done.
Resolving deltas: 100% (146180/146180), done.
```

alvaroaleman commented on prow/git/v2/remote.go:
```diff
@@ -0,0 +1,101 @@
+// (to push to) for a repository. These resolvers
+type RemoteResolverFactory interface {
+	CentralRemote(org, repo string) RemoteResolver
+	PublishRemote(org, repo string) RemoteResolver
```
Nit: Could we have the godocs here? I needed to read them in oder to understand the difference and what consumers will see is this interface, not the implementation, as the latter is not exported

stevekuznetsov commented:
Will move `merge`-related bits over soon, otherwise ready for more looks from @fejta @cjwagner etc -- will work on TODOs today.

stevekuznetsov commented:
@fejta one thought I had was we could also create a new metaclient that had this one and the old one together, then call both at the same time and log differences/errors, allowing us to put this into production but not let it fail and continue to use the old one.

fejta commented:
That sounds good, although I basically think a way to shrink the blast radius is sufficient here. AKA if we can limit it initially to just test infra, prove it works and then use it everywhere I'd be satisfied. Aka if it fails and we just have to fix up test-infra that doesn't seem like a big deal so long as the rest of the org is healthy. 

This is also a nifty idea -- makes it more likely to work when we switch, but doesn't change the blast radius.

petr-muller commented on prow/git/v2/remote.go:
```diff
@@ -0,0 +1,107 @@
+}
+
+// PublishRemote creates a remote resolver that refers to a gitUser's remote
+// for the repository that can be published to.
```
Given I'm now working on something that could use a client like this, but publishing to non-personal, non-central fork. I think we should use some less general name than `PublishRemote` for this piece which can only do personal remotes. I think it's useful to have `PersonalRemote` or `AccountRemote`, let's keep `PublishRemote` to make any remote where an account has write access to.

stevekuznetsov commented:
@cjwagner @fejta @alvaroaleman @petr-muller I've updated this, it is ready. Changes I added:

 - refactor the current `localgit` implementation to return interfaces, not concrete types
 - add adapters so that the v1 types implement the v2 interfaces
 - duplicate all `git` package level integration tests to run with both `v1` and `v2` clients

The `v2` client is mostly a stand-in replacement for the `v1` client -- the only change is accessing the directory is a call to `.Directory()` rather than an access of a public field `.Dir` -- Go would not allow me to have `.Dir()` as well...

stevekuznetsov commented:
Also note to update our cache I use `git remote update` instead  of `git fetch` -- in some cases I was seeing `git fetch` not update symbolic refs like branches, and I could not figure out why. The [doc](https://git-scm.com/docs/git-clone#Documentation/git-clone.txt---mirror) clearly states to use `git remote update` with mirrored clones, so I figure it's fine.

stevekuznetsov commented on prow/git/v2/client_factory.go:
```diff
@@ -0,0 +1,165 @@
+
+// RepoFromDir returns a repository client for a directory that's already initialized with content.
+// If the directory isn't specified, the current working directory is used.
+func (c *clientFactory) RepoFromDir(org, repo, dir string) (RepoClient, error) {
```
Very much scoped to be a "useful thing for Prow interacting with GitHub remotes"

stevekuznetsov commented:
```
ERROR: An error occurred during the fetch of repository 'jsonnet':
   java.io.IOException: Error downloading [https://github.com/google/jsonnet/archive/v0.13.0.tar.gz] to /bazel-scratch/.cache/bazel/_bazel_root/05618a594cb3499a4b912817865867cd/external/jsonnet/v0.13.0.tar.gz: Premature EOF 
```

/retest

matthyx commented:
/uncc

stevekuznetsov commented:
Just when I thought I was done ... I forgot I had to chase the interface through the whole codebase. So, this PR is now even more huge with the latest commit. However, the commit is mechanical and I can break this out into many PRs if people would like. An overview of the current changes:

1. add an interface to close over the git client behavior
2. implement a v2 library
3. add an adapter from v1 to v2, which is trivially simple -- it actually adapts only two methods (one is a change from a field access to a getter, the other is a change to elide defaulted data instead of asking people to send it)
4. refactor the code to expect the interface
5. refactor the setup to continue using the v1 client, but behind the adapter
6. duplicate tests to run with v1 and v2 to show that v2 is functional

There is no behavioral change to Prow here, yet, since everyone is still using the v1 client (albeit behind the adpater). If we want, we can split all of the above into separate PRs.

alvaroaleman commented on prow/git/v2/client_factory.go:
```diff
@@ -0,0 +1,201 @@
+	ClientFor(org, repo string) (RepoClient, error)
+
+	// Clean removes the caches used to generate clients
+	Clean() error
```
Do we have a use-case for this outside of tests?

alvaroaleman commented on prow/cmd/deck/main.go:
```diff
@@ -581,7 +581,7 @@ func prodOnlyMain(cfg config.Getter, pluginAgent *plugins.ConfigAgent, o options
 	// presubmits dynamically which we need for the PR history page.
 	var githubClient deckGitHubClient
-	var gitClient *git.Client
+	var gitClient git.ClientFactory
```
Generally I prefer to have distinct PRs for the implementation and the integration, if possible. It has the advantage that its easier to review and makes me as a reviewer feel less "I really must not miss something". Also, with the combined PR if we have to revert the feature as a whole gets removed again rather than just the integration.

alvaroaleman commented on prow/git/v2/adapter.go:
```diff
@@ -0,0 +1,82 @@
+	return errors.New("no Fetch implementation exists in the v1 repo client")
+}
+
+func (a *repoClientAdapter) RemoteUpdate() error {
```
Why do we do this rather than just letting the compilation fail?

alvaroaleman commented on prow/git/v2/client_factory.go:
```diff
@@ -0,0 +1,201 @@
+
+// NewLocalClientFactory allows for the creation of repository clients
+// based on a local filepath remote for testing
+func NewLocalClientFactory(baseDir string, gitUser func() (name, email string, err error), censor func(content []byte) []byte) (ClientFactory, error) {
```
I already thought that about the initial naming of "localgit": What about renaming this to `NewTestClientFactory` to clarify the purpose? The  `local` part is more of an implementation detail than telling anything about the purpose

alvaroaleman commented on prow/git/v2/executor.go:
```diff
@@ -0,0 +1,72 @@
+	Run(args ...string) ([]byte, error)
+}
+
+func NewCensoringExecutor(dir string, censor func(content []byte) []byte, logger *logrus.Entry) (Executor, error) {
```
Are we planning on ever adding a non-censoring executor? If not, maybe rename to just `NewExecutor`?

stevekuznetsov commented:
I cannot reproduce that test ... what is going on?

```
$ bazel test //...
/* holy crap lots of output */
@io_k8s_repo_infra//hack:verify-gofmt                                    PASSED in 2.3s

Executed 77 out of 208 tests: 208 tests pass.
INFO: Build completed successfully, 2089 total actions
```

stevekuznetsov commented:
```
$ bazel build //prow/cmd/tide/...
INFO: Build option --features has changed, discarding analysis cache.
INFO: Analyzed 8 targets (28 packages loaded, 10421 targets configured).
INFO: Found 8 targets...
INFO: Elapsed time: 45.102s, Critical Path: 19.23s
INFO: 520 processes: 520 linux-sandbox.
INFO: Build completed successfully, 521 total actions
```

alvaroaleman commented on prow/git/v2/interactor.go:
```diff
@@ -0,0 +1,279 @@
+	"os"
+)
+
+// Interactor knows how to publish local work to a remote
```
I think `interactor` is just for local interaction not for pushing  or not?

alvaroaleman commented on prow/git/v2/interactor.go:
```diff
@@ -0,0 +1,279 @@
+// by performing a three-way merge (similar to git cherry-pick). It returns
+// an error if the patch cannot be applied.
+func (i *interactor) Am(path string) error {
+	i.logger.Infof("Applying patch at %s", path)
```
IMHO this and pretty much all other logging here should IMHO be debug level, not info

alvaroaleman commented on prow/git/v2/executor.go:
```diff
@@ -0,0 +1,72 @@
+)
+
+// Executor knows how to execute Git commands
+type Executor interface {
```
Is this really an interface we want to expose?

stevekuznetsov commented:
Closing in favor of 
#15362 
#15362 
#15362 
#15362

stevekuznetsov commented:
/close
